### DISCUSSION TITLE
Layer structuring

### DISCUSSION THREAD

**Discussion by Gerry546** [meta-homeassistant/meta-homeassistant#76] [Ideas]
I personally have a lot of trouble to keep track of all the individual recipes from python which are required as dependencies for HA to run. 

Now python recipes are placed in two folders in the layer:
- recipes-devtools/python
- recipes-homeassistant/homeassistant
 
My suggestion is to structure this a bit different. Also to make it more easy to maintain the layer since it will be easier for people to understand the structuring of the layer.

My suggestion would be:
- recipes-homeassistant/homeassistant: 
  - for the recipe of home assistant itself
- recipes-homeassistant/homeassistant-core-deps
  - there are several dependency recipes which are directly needed to be able to even boot HA. I would put them separate since they are also likely to change in version.
- recipes-homeassistant/homeassistant-component-deps
  -  most HA components actually use very specifically developed python modules, and are also prone to frequent updates.
- recipes-devtools/python:
  - there are multiple python recipes which, in turn, have dependencies on other python recipes. Not all of these are provided in OE or Poky. I would place these secondary recipes in the regular layer ordering.

**Comment by bachp** [comment]
I fully support this idea, I think the new layering makes sense.

**Comment by Gerry546** [comment]
As I'm preparing for the update to december 2023 I would like to revisit this discussion.
I have worked with my previously proposed structuring but I have noticed that it has not achieved the effect that I would like. From one side due to the fact that dependencies pop up, or dissapear quite fast so having them separated into the layer structure before was making it harder for me to maintain instead of easier.

### Layer structure
So I would like to refine the structuring of the layer.
For starters I would like to propose the following:
![image](https://github.com/meta-homeassistant/meta-homeassistant/assets/59296539/20fe5a13-2fbf-408a-b337-2ed6385d6d86)

What we have is:
- recipes-homeassistant: Everything directly related to HomeAssistant
- home-assistant-libs: All python dependencies hosted by the github: https://github.com/home-assistant-libs
- homeassistant: All python dependencies hosted by the github: https://github.com/home-assistant
- images: the demo images to be with homeassistant included.
- nabucasa: the cloud platform for HA.
- and essentially everything else back into recipes-devtools/python.

### python3-homeassistant structure
A small second topic I would like to address is the recipe for `python3-homeassistant` itself. I want to start expanding the list of components which are supported. However keeping everything into one big recipe file will become unwieldy. What I would like to do is to take the components out mostly and put them into `.inc` files.
What I would like to do is to make one `.inc` file per integration 'category' in a similar fashion as homeassistant itself also divides them: https://www.home-assistant.io/integrations/

Let me know what you think of it.

**Reply by bachp** [reply]
Your proposal makes sense, no objection from my side.

**Reply by Gerry546** [reply]
I am testing it out in my upcoming change for 2023.12.0.

What I'm unsure about is about integrations which are listed in multiple categories.
I don't want to duplicate code so I have two options I think:
- Put them in a `multiple.inc`.
- Put them in a single (arbitrarily chosen) `inc` file.

Maybe you guys have an alternative?
I put the it up in draft for merge here #112  so you can have a look.
